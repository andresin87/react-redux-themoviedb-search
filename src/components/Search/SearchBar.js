import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Input } from 'antd';

import './SearchBar.css';
import { fetchSearch } from '../../actions';

const Search = Input.Search;

class SearchBar extends PureComponent {
  render() {
    const { fetchSearch } = this.props;
    return (
      <div className="SearchBar">
        <Search
          placeholder="Search for a film..."
          enterButton="Search"
          size="large"
          onSearch={value => fetchSearch(value)}
        />
      </div>
    );
  }
}

SearchBar.propTypes = {
  fetchSearch: PropTypes.func.isRequired,
};

export default connect(
  null,
  { fetchSearch }
)(SearchBar);
