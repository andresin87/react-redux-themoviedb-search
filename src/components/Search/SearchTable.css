.SearchTable {
  padding: 5px;
  margin: 10px;
  border-radius: 8px;
}

@media (max-width: 1400px) {
  .SearchTableGenres {
    display: none;
  }
}

@media (max-width: 1100px) {
  .SearchTableReleaseDate {
    display: none;
  }
}

@media (max-width: 950px) {
  .SearchTablePopularity {
    display: none;
  }
}

@media (max-width: 700px) {
  .SearchTableVoteCount {
    display: none;
  }
}

@media (max-width: 550px) {
  .SearchTableVoteAverage {
    display: none;
  }
}
