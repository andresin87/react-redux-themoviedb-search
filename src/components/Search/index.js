import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import { fetchPopular, fetchSearch } from '../../actions';

class Search extends PureComponent {
	componentDidMount() {
		const { fetchPopular } = this.props;
		fetchPopular();
	}
	render() {
		const { search } = this.props;
		return (
			<pre style={{ textAlign: 'left' }}>{JSON.stringify(search, null, 2)}</pre>
		);
	}
}

const mapStateToProps = state => {
	const { search } = state;
	return { search };
};

export default connect(
	mapStateToProps,
	{ fetchPopular, fetchSearch }
)(Search);
