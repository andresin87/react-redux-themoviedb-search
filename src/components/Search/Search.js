import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import SearchBar from './SearchBar';
import SearchTable from './SearchTable';
import { selectorSearch } from '../../reducers/search';
import { selectorGenres } from '../../reducers/genres';
import { fetchPopular, fetchGenres } from '../../actions';

class Search extends PureComponent {
  componentDidMount() {
    const {
      isLoadingMovies,
      movies,
      isLoadingGenres,
      genresNames,
    } = this.props;
    const { fetchPopular, fetchGenres } = this.props;
    if (!isLoadingMovies && movies.length === 0) fetchPopular();
    if (!isLoadingGenres && Object.keys(genresNames).length === 0)
      fetchGenres();
  }
  render() {
    return (
      <Fragment>
        <SearchBar />
        <SearchTable />
      </Fragment>
    );
  }
}

Search.propTypes = {
  movies: PropTypes.array.isRequired,
  isLoadingMovies: PropTypes.bool.isRequired,
  genresNames: PropTypes.object.isRequired,
  isLoadingGenres: PropTypes.bool.isRequired,
  fetchPopular: PropTypes.func.isRequired,
  fetchGenres: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  const { movies, isLoading: isLoadingMovies } = selectorSearch(state);
  const { names: genresNames, isLoading: isLoadingGenres } = selectorGenres(
    state
  );
  return {
    movies,
    isLoadingMovies,
    genresNames,
    isLoadingGenres,
  };
};

export default connect(
  mapStateToProps,
  { fetchPopular, fetchGenres }
)(Search);
