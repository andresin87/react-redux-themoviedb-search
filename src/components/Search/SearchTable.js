import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Table } from 'antd';
import moment from 'moment';

import './SearchTable.css';
import { selectorSearch } from '../../reducers/search';
import { selectorGenres } from '../../reducers/genres';

const sort = (itemA, itemB) => {
  if (itemA > itemB) return 1;
  if (itemA < itemB) return -1;
  return 0;
};

class SearchTable extends PureComponent {
  static contextTypes = {
    router: PropTypes.object,
  };

  goToDetail(id, dataSource) {
    this.context.router.history.push({
      pathname: `/detail/${id}`,
      state: { dataSource },
    });
  }

  getColumns = () => {
    const { genresNames } = this.props;
    const columns = [
      {
        title: 'Title',
        dataIndex: 'title',
        key: 'title',
        className: 'SearchTableTitle',
        sorter: (a, b) => sort(a.title, b.title),
        width: 450,
      },
      {
        title: 'Vote average',
        dataIndex: 'vote_average',
        key: 'vote_average',
        className: 'SearchTableVoteAverage',
        sorter: (a, b) => sort(a.vote_average, b.vote_average),
        width: 150,
        align: 'center',
      },
      {
        title: 'Vote count',
        dataIndex: 'vote_count',
        key: 'vote_count',
        className: 'SearchTableVoteCount',
        sorter: (a, b) => sort(a.vote_count, b.vote_count),
        width: 150,
        align: 'center',
      },
      {
        title: 'Popularity',
        dataIndex: 'popularity',
        key: 'popularity',
        className: 'SearchTablePopularity',
        sorter: (a, b) => sort(a.popularity, b.popularity),
        width: 150,
        align: 'center',
      },
      {
        title: 'Release date',
        dataIndex: 'release_date',
        key: 'release_date',
        className: 'SearchTableReleaseDate',
        render: text => moment(text).format('DD/MM/YYYY'),
        sorter: (a, b) => sort(a.release_date, b.release_date),
        width: 150,
        align: 'center',
      },
      {
        title: 'Genres',
        dataIndex: 'genre_ids',
        key: 'genre_ids',
        render: ids =>
          ids
            .reduce((genresAsString, idGenre) => {
              return genresNames && genresNames[idGenre]
                ? `${genresAsString} ${genresNames[idGenre]},`
                : '';
            }, '')
            .slice(0, -1),
        className: 'SearchTableGenres',
      },
    ];
    return columns;
  };

  render() {
    const { isLoadingMovies, movies, isLoadingGenres } = this.props;
    const columns = this.getColumns();
    return (
      <div className="SearchTable">
        <Table
          loading={isLoadingMovies || isLoadingGenres}
          dataSource={movies}
          columns={columns}
          pagination={{ pageSize: 10 }}
          onRow={record => {
            return {
              onClick: () => this.goToDetail(record.id, movies),
            };
          }}
          rowKey={record => record.id}
        />
      </div>
    );
  }
}

SearchTable.propTypes = {
  movies: PropTypes.array.isRequired,
  isLoadingMovies: PropTypes.bool.isRequired,
  genresNames: PropTypes.object.isRequired,
  isLoadingGenres: PropTypes.bool.isRequired,
};

const mapStateToProps = state => {
  const { movies, isLoading: isLoadingMovies } = selectorSearch(state);
  const { names: genresNames, isLoading: isLoadingGenres } = selectorGenres(
    state
  );
  return {
    movies,
    isLoadingMovies,
    genresNames,
    isLoadingGenres,
  };
};

export default connect(mapStateToProps)(SearchTable);
