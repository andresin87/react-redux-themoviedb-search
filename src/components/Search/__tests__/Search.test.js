import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Search from '../Search';
import store from '../../../config/mockStore';
import { fetchPopular, fetchGenres } from '../../../actions';

jest.mock('../../../actions');

describe('Search component', () => {
  const initialEntries = ['/'];
  const component = (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>
        <Search />
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    fetchPopular.mockReturnValue({ type: 'fetchPopular' });
    fetchGenres.mockReturnValue({ type: 'fetchGenres' });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Snapshot matchs', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('First mount: we must load the popular films', () => {
    const middlewares = [thunk];
    const mockStore = configureMockStore(middlewares);

    const store = mockStore({
      search: { movies: [], isLoading: false },
      genres: { names: {}, isLoading: false },
    });
    const component = (
      <Provider store={store}>
        <MemoryRouter initialEntries={initialEntries}>
          <Search />
        </MemoryRouter>
      </Provider>
    );

    it('Calls to fetchPopular on load', () => {
      const div = document.createElement('div');
      ReactDOM.render(component, div);
      expect(fetchPopular).toHaveBeenCalledTimes(1);
      expect(fetchPopular).toHaveBeenLastCalledWith();
    });

    it('Calls to fetchGenres on load', () => {
      const div = document.createElement('div');
      ReactDOM.render(component, div);
      expect(fetchGenres).toHaveBeenCalledTimes(1);
      expect(fetchGenres).toHaveBeenLastCalledWith();
    });
  });

  describe('Next mounts: we must mantain the current data', () => {
    it(`Doesn't call to fetchPopular on load`, () => {
      const div = document.createElement('div');
      ReactDOM.render(component, div);
      expect(fetchPopular).toHaveBeenCalledTimes(0);
    });

    it(`Doesn't call to fetchGenres on load`, () => {
      const div = document.createElement('div');
      ReactDOM.render(component, div);
      expect(fetchGenres).toHaveBeenCalledTimes(0);
    });
  });
});
