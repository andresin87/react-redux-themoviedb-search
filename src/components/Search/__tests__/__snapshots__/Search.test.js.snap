// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Snapshot matchs 1`] = `
Array [
  <code
    name="SearchBar"
    props={Object {}}
  />,
  <code
    name="SearchTable"
    props={Object {}}
  />,
]
`;
